### PR TITLE
fix: Dovecot Quota dummy accounts for aliases should check for existing users with an exact user key lookup

### LINKED ISSUE #2639: [BUG] Alias will not be added to '/etc/dovecot/userdb' twice
### Miscellaneous first checks

- [X] I checked that all ports are open and not blocked by my ISP / hosting provider.
- [X] I know that SSL errors are likely the result of a wrong setup on the user side and not caused by DMS itself. I'm confident my setup is correct.

### Affected Component(s)

Dovecot

### What happened and when does this occur?

The Dovecot regexp catches false positive.

```Markdown
mailsserver | [  DEBUG  ]  Adding alias 'prefixtest@domain.tld' for user 'user@domain.tld' to Dovecot's userdb
mailsserver | [  DEBUG  ]  Adding alias 'test@domain.tld' for user 'user@domain.tld' to Dovecot's userdb
mailsserver | [ WARNING ]  Alias 'test@domain.tld' will not be added to '/etc/dovecot/userdb' twice
```

### What did you expect to happen?

Expecting alias to be added being different than the first alias provided.

### How do we replicate the issue?

```Markdown
Tested in a fresh install with latest version (11.0.0)

1. Add these 2 aliases in `postfix-virtual.cf`

prefixtest@domain.tld user@domain.tld
test@domain.tld user@domain.tld

2. start server in debug mode
3. you should see the error listed above

Note:
* the bug happens ONLY when the longer alias is set first in the postfix-virtual.cf
```

### DMS version

v11.0.0

### What operating system is DMS running on?

Linux

### What instruction set architecture is DMS running on?

x86_64 / AMD64

### What container orchestration tool are you using?

Docker Compose

### docker-compose.yml

```yml
Not related.
```

### Relevant log output

```bash
Full log:

mailsserver | [   INF   ]  Welcome to docker-mailserver 11.0.0
mailsserver | [   INF   ]  Initializing setup
mailsserver | [  DEBUG  ]  Registering functions
mailsserver | [   INF   ]  Checking configuration
mailsserver | [  DEBUG  ]  Checking that hostname/domainname is provided or overridden
mailsserver | [  DEBUG  ]  Domain has been set to domain.tld
mailsserver | [  DEBUG  ]  Hostname has been set to mail.domain.tld
mailsserver | [   INF   ]  Configuring mail server
mailsserver | [  DEBUG  ]  Setting up default variables
mailsserver | [  DEBUG  ]  Setting file and directory permissions
mailsserver | [  DEBUG  ]  Setting up Dovecot
mailsserver | [  DEBUG  ]  Spam messages will be moved to the Junk folder
mailsserver | [  DEBUG  ]  Setting up Dovecot dhparam
mailsserver | [  DEBUG  ]  Setting up Dovecot quota
mailsserver | [  DEBUG  ]  Setting up Dovecot Local User
mailsserver | [  DEBUG  ]  Creating user 'user' for domain 'domain.tld'
mailsserver | [  DEBUG  ]  Adding alias 'prefixtest@domain.tld' for user 'user@domain.tld' to Dovecot's userdb
mailsserver | [  DEBUG  ]  Adding alias 'test@domain.tld' for user 'user@domain.tld' to Dovecot's userdb
mailsserver | [ WARNING ]  Alias 'test@domain.tld' will not be added to '/etc/dovecot/userdb' twice
mailsserver | [  DEBUG  ]  Configuring Postgrey
mailsserver | [  DEBUG  ]  Setting up Fail2Ban
mailsserver | [  DEBUG  ]  Disabling postfix DNS block list (zen.spamhaus.org)
mailsserver | [  DEBUG  ]  Disabling postscreen DNS block lists
mailsserver | [  DEBUG  ]  Setting up DKIM
mailsserver | [ WARNING ]  No DKIM key(s) provided - check the documentation on how to get your keys
mailsserver | Nameservers 127.0.0.11
mailsserver | [  DEBUG  ]  Setting up SSL
mailsserver | [  DEBUG  ]  TLS configured with 'modern' ciphers
mailsserver | [ WARNING ]  !! INSECURE !! SSL configured with plain text access - DO NOT USE FOR PRODUCTION DEPLOYMENT
mailsserver | [  DEBUG  ]  Setting up PERMIT_DOCKER option
mailsserver | [  DEBUG  ]  Setting up mailname and creating '/etc/mailname'
mailsserver | [  DEBUG  ]  Setting up Amavis
mailsserver | [  DEBUG  ]  Setting up DMARC
mailsserver | [  DEBUG  ]  Applying hostname and domainname to Postfix
mailsserver | [  DEBUG  ]  Applying hostname to Dovecot
mailsserver | [  DEBUG  ]  Setting up Postfix SASL Password
mailsserver | [  DEBUG  ]  SASL_PASSWD was not provided - '/etc/postfix/sasl_passwd' not created
mailsserver | [  DEBUG  ]  Setting up Security Stack
mailsserver | [  DEBUG  ]  Enabling and configuring SpamAssassin
mailsserver | [  DEBUG  ]  Enabling ClamAV
mailsserver | [  DEBUG  ]  Enabling Fail2Ban
mailsserver | [  DEBUG  ]  Enabling Amavis
mailsserver | [  DEBUG  ]  Setting up Postfix aliases
mailsserver | [  DEBUG  ]  Setting up Postfix vhost
mailsserver | [  DEBUG  ]  Setting up Postfix dhparam
mailsserver | [  DEBUG  ]  Configuring Postscreen
mailsserver | [  DEBUG  ]  Setting up Postfix Relay Hosts
mailsserver | [  DEBUG  ]  Setting up logrotate
mailsserver | [  DEBUG  ]  Postfix log summary reports disabled
mailsserver | [  DEBUG  ]  Logwatch reports disabled.
mailsserver | [  DEBUG  ]  Setting up configuration checksum file
mailsserver | [  DEBUG  ]  Printing environment variables. Make sure no sensitive data is copied.
mailsserver | AMAVIS_LOGLEVEL='0'
mailsserver | CLAMAV_MESSAGE_SIZE_LIMIT='25M'
mailsserver | DEFAULT_RELAY_HOST=''
mailsserver | DOVECOT_INET_PROTOCOLS='all'
mailsserver | DOVECOT_MAILBOX_FORMAT='maildir'
mailsserver | DOVECOT_TLS='no'
mailsserver | ENABLE_AMAVIS='1'
mailsserver | ENABLE_CLAMAV='1'
mailsserver | ENABLE_DNSBL='0'
mailsserver | ENABLE_FAIL2BAN='1'
mailsserver | ENABLE_FETCHMAIL='0'
mailsserver | ENABLE_LDAP='0'
mailsserver | ENABLE_MANAGESIEVE='0'
mailsserver | ENABLE_POP3='0'
mailsserver | ENABLE_POSTGREY='1'
mailsserver | ENABLE_QUOTAS='1'
mailsserver | ENABLE_SASLAUTHD='0'
mailsserver | ENABLE_SPAMASSASSIN='1'
mailsserver | ENABLE_SPAMASSASSIN_KAM='0'
mailsserver | ENABLE_SRS='0'
mailsserver | ENABLE_UPDATE_CHECK='1'
mailsserver | FAIL2BAN_BLOCKTYPE='drop'
mailsserver | FETCHMAIL_PARALLEL='0'
mailsserver | FETCHMAIL_POLL='300'
mailsserver | LDAP_START_TLS='no'
mailsserver | LOGROTATE_INTERVAL='weekly'
mailsserver | LOGWATCH_INTERVAL='none'
mailsserver | LOGWATCH_RECIPIENT='postmaster@domain.tld'
mailsserver | LOGWATCH_SENDER='mailserver-report@mail.domain.tld'
mailsserver | LOG_LEVEL='debug'
mailsserver | MOVE_SPAM_TO_JUNK='1'
mailsserver | NETWORK_INTERFACE='eth0'
mailsserver | ONE_DIR='1'
mailsserver | OVERRIDE_HOSTNAME=''
mailsserver | PERMIT_DOCKER='none'
mailsserver | PFLOGSUMM_RECIPIENT='postmaster@domain.tld'
mailsserver | PFLOGSUMM_SENDER='mailserver-report@mail.domain.tld'
mailsserver | PFLOGSUMM_TRIGGER='none'
mailsserver | POSTFIX_INET_PROTOCOLS='all'
mailsserver | POSTFIX_MAILBOX_SIZE_LIMIT='0'
mailsserver | POSTFIX_MESSAGE_SIZE_LIMIT='10240000'
mailsserver | POSTGREY_AUTO_WHITELIST_CLIENTS='5'
mailsserver | POSTGREY_DELAY='300'
mailsserver | POSTGREY_MAX_AGE='35'
mailsserver | POSTGREY_TEXT='Delayed by Postgrey'
mailsserver | POSTMASTER_ADDRESS='postmaster@domain.tld'
mailsserver | POSTSCREEN_ACTION='enforce'
mailsserver | RELAY_HOST=''
mailsserver | REPORT_RECIPIENT='postmaster@domain.tld'
mailsserver | REPORT_SENDER='mailserver-report@mail.domain.tld'
mailsserver | SA_KILL='6.31'
mailsserver | SA_SPAM_SUBJECT='***SPAM*** '
mailsserver | SA_TAG2='6.31'
mailsserver | SA_TAG='2.0'
mailsserver | SMTP_ONLY='0'
mailsserver | SPAMASSASSIN_SPAM_TO_INBOX='1'
mailsserver | SPOOF_PROTECTION='0'
mailsserver | SRS_SENDER_CLASSES='envelope_sender'
mailsserver | SSL_TYPE=''
mailsserver | SUPERVISOR_LOGLEVEL='warn'
mailsserver | TLS_LEVEL='modern'
mailsserver | TZ=''
mailsserver | UPDATE_CHECK_INTERVAL='1d'
mailsserver | VIRUSMAILS_DELETE_DELAY='7'
mailsserver | [   INF   ]  Post-configuration checks
mailsserver | [  DEBUG  ]  Checking /var/mail permissions
mailsserver | [   INF   ]  Starting miscellaneous tasks
mailsserver | [  DEBUG  ]  Consolidating all state onto /var/mail-state
mailsserver | [   INF   ]  Starting daemons
mailsserver | [  DEBUG  ]  Starting cron
mailsserver | [  DEBUG  ]  Starting rsyslog
mailsserver | [  DEBUG  ]  Starting dovecot
mailsserver | [  DEBUG  ]  Starting update-check
mailsserver | [  DEBUG  ]  Starting opendkim
mailsserver | [  DEBUG  ]  Starting opendmarc
mailsserver | [  DEBUG  ]  Starting postgrey
mailsserver | [  DEBUG  ]  Starting postfix
mailsserver | [  DEBUG  ]  Starting fail2ban
mailsserver | [  DEBUG  ]  Starting clamav
mailsserver | [  DEBUG  ]  Starting changedetector
mailsserver | [  DEBUG  ]  Starting amavis
mailsserver | [   INF   ]  mail.domain.tld is up and running
mailsserver | Jun 14 10:58:58 mail amavis[798]: starting. /usr/sbin/amavisd-new at mail.domain.tld amavisd-new-2.11.1 (20181009), Unicode aware, LC_CTYPE="C.UTF-8"
mailsserver | Jun 14 10:58:58 mail amavis[798]: perl=5.032001, user=, EUID: 112 (112);  group=, EGID: 114 114 (114 114)
mailsserver | Jun 14 10:58:58 mail amavis[798]: Net::Server: Group Not Defined.  Defaulting to EGID '114 114'
mailsserver | Jun 14 10:58:58 mail amavis[798]: Net::Server: User Not Defined.  Defaulting to EUID '112'
mailsserver | Jun 14 10:58:58 mail amavis[798]: No ext program for   .zoo, tried: zoo
mailsserver | Jun 14 10:58:58 mail amavis[798]: No ext program for   .doc, tried: ripole
mailsserver | Jun 14 10:58:58 mail amavis[798]: No decoder for       .F
mailsserver | Jun 14 10:58:58 mail amavis[798]: No decoder for       .doc
mailsserver | Jun 14 10:58:58 mail amavis[798]: No decoder for       .zoo
mailsserver | Jun 14 10:58:58 mail amavis[798]: Using primary internal av scanner code for ClamAV-clamd
mailsserver | Jun 14 10:58:58 mail amavis[798]: Found secondary av scanner ClamAV-clamscan at /usr/bin/clamscan
mailsserver | Jun 14 10:59:15 mail postfix/master[2004]: daemon started -- version 3.5.6, configuration /etc/postfix
```

### Other relevant information

_No response_

### What level of experience do you have with Docker and mail servers?

- [ ] I am inexperienced with docker
- [ ] I am inexperienced with mail servers
- [ ] I am uncomfortable with the CLI

### Code of conduct

- [X] I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree
- [X] I have read the [README](https://github.com/docker-mailserver/docker-mailserver/blob/master/README.md) and the [documentation](https://docker-mailserver.github.io/docker-mailserver/edge/) and I searched the [issue tracker](https://github.com/docker-mailserver/docker-mailserver/issues?q=is%3Aissue) but could not find a solution

### Improvements to this form?

_No response_

### PR DESCRIPTION
**Pull request by tomav**
# Description

Fixes #2639

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [ ] If necessary I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: target/scripts/helpers/accounts.sh [hunk 1]
```diff
@@ -157,7 +157,7 @@ function _create_dovecot_alias_dummy_accounts
       fi
 
       DOVECOT_USERDB_LINE="${ALIAS}:${REAL_ACC[1]}:5000:5000::/var/mail/${REAL_DOMAINNAME}/${REAL_USERNAME}::${REAL_ACC[2]:-}"
-      if grep -qF "${DOVECOT_USERDB_LINE}" "${DOVECOT_USERDB_FILE}"
```
Comment (georglauterbach) [inline target/scripts/helpers/accounts.sh:160]: Why was the `-F` option removed? IIRC @casperklein mentioned (please correct me if I‘m wrong), that we need to search for the pattern literally, i.e. with fixed strings. If this is indeed not needed, all is well :D

**Comment by georglauterbach** [comment]
@polarathene does this already suffice when it comes to the issue of adding „wildcard“ aliases?

File: target/scripts/helpers/accounts.sh (hunk 1 above)
Comment (casperklein) [reply target/scripts/helpers/accounts.sh:160]: `-F` is useful/best practice, when no regex is used and just plain text should be matched. So when adding `^`, `-F` must be removed 👍

**Review by casperklein — APPROVED** [review]

File: target/scripts/helpers/accounts.sh (hunk 1 above)
Comment (georglauterbach) [reply target/scripts/helpers/accounts.sh:160]: Ah yes, right. Would it make sense then 
to add a `$` at the end here too?

Comment (polarathene) [reply target/scripts/helpers/accounts.sh:160]: If you want to match the whole line exactly, `grep -qxF` would work (same as if we additionally had `^`+`$` tacked on both ends in a compatible manner.

However this specific check is only interested in the duplicate user entry really, so we don't need to match the entire line.. just the first column. A key lookup such as `grep -qi "^${ALIAS}:" "${DOVECOT_USERDB_FILE}"` would suffice.

File: target/scripts/helpers/accounts.sh [hunk 2]
```diff
@@ -157,7 +157,7 @@ function _create_dovecot_alias_dummy_accounts
...
+      if grep -q "^${DOVECOT_USERDB_LINE}" "${DOVECOT_USERDB_FILE}"
```
Comment (polarathene) [inline target/scripts/helpers/accounts.sh:160]:
```suggestion
      if grep -qi "^${ALIAS}:" "${DOVECOT_USERDB_FILE}"
```

**Review by polarathene — CHANGES_REQUESTED** [review]
Only match what's relevant, the user column. Ensure the value is the start of the line and ends at the delimiter `:` to avoid the current bug.

Additionally I don't think we want to accidentally have two users with different casing, so use case insensitive match `-i`.

**Comment by polarathene** [comment]
> @polarathene does this already suffice when it comes to the issue of adding „wildcard“ aliases ([#2637 (comment)](https://github.com/docker-mailserver/docker-mailserver/issues/2637#issuecomment-1153120021))?

No. That is a separate issue.

If I understand correctly, the method would still accept a `@example.com` wildcard alias entry as a user and enter it into the user DB.

You would have to reject an address that starts with `@` with `^@`(identify wildcard) or `^[^@]` (identify valid entry as any first char that is not `@`).

---

That said, this methods purpose only is as an awful workaround for supporting `ENABLE_QUOTAS=1`. Technically without adding the wildcard entry, you'd get the problem with the postfix policy check as no alias user `@example.com` exists.

In that sense, by not addressing it like I mentioned, yes, this PR should also fix that issue. It may not fix it fully for that user as they had recursive aliases (aliases that referenced aliases). A similar issue may be present with support for multiple recipients to an alias, as I don't think the current code handles that either.

Unless someone adds a proper policy service to resolve aliases to actual users, the quota support is going to be flakey. AFAIK even then, I think it would be an issue with multiple recipients, as you either accept or reject when at least one alias mapped user has a known quota overage. It'd seem wrong to block delivery to other recipients for the alias, but not doing so would defeat the purpose afaik regarding backscatter prevention.

File: target/scripts/helpers/accounts.sh (hunk 2 above)
Comment (georglauterbach) [reply target/scripts/helpers/accounts.sh:160]: 👍🏼

**Review by polarathene — APPROVED** [review]

**Comment by polarathene** [comment]
We can squeeze in this fix for 11.1 I think.

**Review by georglauterbach — APPROVED** [review]
Yes, this is fine to be in 11.1.0 :)

**Review by casperklein — APPROVED** [review]

**Comment by tomav** [comment]
Well, thanks guys.

**Pull request merged**